rfcs/0000-graphql-api.md

```diff
@@ -0,0 +1,433 @@
+just to avoid needing to query it all every time.
+
+Note: There are several custom scalars defined at the top for readability. These will be `String`s in the final implementation due to the complications involved in the encoding of custom scalars not being expressed in the schema and needing to be implemented symmetrically on the client and server.
+
+Note: Public keys will all be the "compressed" public keys that are used elsewhere in the node.
+
+```graphql
+# Note: this will all be strings in the actual api.
+scalar Date
+scalar PublicKey
+scalar PrivateKey
+scalar UInt64
+scalar UInt32
+
+enum ConsensusStatus {
+  SUBMITTED
+  INCLUDED # Included in any block
+  FINALIZED
+  SNARKED
+  FAILED
+}
+
+type ConsensusState {
+  status: ConsensusStatus!
+  estimatedPercentConfirmed: Float!
+}
+
+type Payment {
+  nonce: Int!,
+  submittedAt: Date!
+  includedAt: Date
+  from: PublicKey! 
+  to: PublicKey!
+  amount: UInt64!
+  fee: UInt32!,
+  memo: String
+}
+
+type PaymentUpdate {
+  payment: Payment
+  consensus: ConsensusState!
+}
+
+enum SyncStatus {
+  ERROR
+  BOOTSTRAP # Resyncing
+  STALE # You haven't seen any activity recently
+  SYNCED
+}
+
+type SyncUpdate {
+  status: SyncStatus!
+  estimatedPercentSynced: Float!
+  description: String
+}
+
+type SnarkWorker {
+  key: PublicKey!
+  fee: UInt32!
+}
+
+type SnarkFee {
+  snarkCreator: PublicKey!
+  fee: UInt32!
+}
+
+type SnarkFeeUpdate {
+  fee: SnarkFee
+  consensus: ConsensusState!
+}
+
+type Block {
+  coinbase: UInt32!
+  creator: PublicKey!
+  payments: [Payment]!
+  snarkFees: [SnarkFee]!
+}
+
+type BlockUpdate {
+  block: Block!
+  consensus: ConsensusState!
+}
+
+type Balance {
+  total: UInt64!,
+  unknown: UInt64!,
+}
+
+type Wallet {
+  publicKey: PublicKey!
+  balance: Balance!
+}
+
+type NodeStatus {
+  network: String
+}
+
+## Input types
+
+input AddWalletInput {
+  public: PublicKey
+  private: PrivateKey
+}
+
+input DeleteWalletInput {
+  public: PublicKey
+}
+
+input AddPaymentReceiptInput {
+  receipt: String
+}
+
+input SetNetworkInput {
+  address: String
+}
+
+input SetSnarkWorkerInput {
+  worker: PublicKey!
+  fee: UInt32!
+}
+
+input CreatePaymentInput {
+  from: PublicKey!,
+  to: PublicKey!,
+  amount: UInt64!,
+  fee: UInt32!,
+  memo: String
+}
+
+input PaymentFilterInput {
+  toOrFrom: PublicKey,
+}
+
+input BlockFilterInput {
+  creator: PublicKey,
+}
+
+## Payload types
+
+type CreatePaymentPayload {
+  payment: Payment
+}
+
+type SetSnarkWorkerPayload {
+  worker: SnarkWorker
+}
+
+type SetNetworkPayload {
+  address: String
+}
+
+type AddPaymentReceiptPayload {
+  payment: Payment
+}
+
+type AddWalletPayload {
+  publicKey: PublicKey
+}
+
+type DeleteWalletPayload {
+  publicKey: PublicKey
+}
+
+# Pagination types
+
+type PageInfo {
+  hasPreviousPage: Boolean!
+  hasNextPage: Boolean!
+}
+
+type PaymentEdge {
+  cursor: String
+  node: PaymentUpdate
+}
+
+type PaymentConnection {
+  edges: [PaymentEdge]
+  nodes: [PaymentUpdate]
```

Comment:
I am pretty sure that a `node` is contained in an `edge` in Relay. So I am going to omit this field.

Reply:
`nodes` is redundant to `edges`, yeah. I believe the idea is to make it easier to casually query/inspect the data without needing to deal with pagination.
Probably no harm in leaving it out for now.